src/lib.rs

```diff
@@ -400,31 +425,35 @@ impl Wal {
             return Ok(());
         }
 
-        // If `until` goes into or above our open segment, delete all but the last closed segments
+        let preserved_closed_start_index = self.preserved_closed_start_index();
+
+        // If `until` goes into or above our open segment, delete till preserved closed segment index
         if until >= self.open_segment_start_index() {
-            for segment in self
-                .closed_segments
-                .drain(..self.closed_segments.len().saturating_sub(1))
-            {
+            for segment in self.closed_segments.drain(..preserved_closed_start_index) {
+                // todo: verify
                 segment.segment.delete()?
             }
             return Ok(());
         }
 
         // Delete all closed segments before the one `until` is in
         let index = self.find_closed_segment(until).unwrap();
-        trace!("{self:?}: prefix truncating until segment {index}");
+        let index = index.min(preserved_closed_start_index);
+        trace!("{self:?}: prefix truncating until closed segment {index}");
```

Comment:
Let's keep the range within the `trace!()` call so that we skip it when we don't have trace logging enabled

Reply:
didn't get you 🤔

Reply:
But we need to use this new index value inside self.closed_segments.drain() in the next line.

Reply:
You are completely right, missed that. Thanks!